datafusion/functions-aggregate/src/median.rs

```diff
@@ -230,6 +276,201 @@ impl<T: ArrowNumericType> Accumulator for MedianAccumulator<T> {
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 1, "single argument to update_batch");
+        let values = values[0].as_primitive::<T>();
+
+        // Push the `not nulls + not filtered` row into its group
+        self.group_values.resize(total_num_groups, Vec::new());
+        accumulate(
+            group_indices,
+            values,
+            opt_filter,
+            |group_index, new_value| {
+                self.group_values[group_index].push(new_value);
+            },
+        );
+
+        Ok(())
+    }
+
+    fn merge_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indices: &[usize],
+        // Since aggregate filter should be applied in partial stage, in final stage there should be no filter
+        _opt_filter: Option<&BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 1, "one argument to merge_batch");
+
+        // The merged values should be organized like as a `ListArray` which is nullable
+        // (input with nulls usually generated from `convert_to_state`), but `inner array` of
+        // `ListArray`  is `non-nullable`.
+        //
+        // Following is the possible and impossible input `values`:
+        //
+        // # Possible values
+        // ```text
+        //   group 0: [1, 2, 3]
+        //   group 1: null (list array is nullable)
+        //   group 2: [6, 7, 8]
+        //   ...
+        //   group n: [...]
+        // ```
+        //
+        // # Impossible values
+        // ```text
+        //   group x: [1, 2, null] (values in list array is non-nullable)
+        // ```
+        //
+        let input_group_values = values[0].as_list::<i32>();
+
+        // Ensure group values big enough
+        self.group_values.resize(total_num_groups, Vec::new());
+
+        // Extend values to related groups
+        // TODO: avoid using iterator of the `ListArray`, this will lead to
+        // many calls of `slice` of its ``inner array`, and `slice` is not
+        // so efficient(due to the calculation of `null_count` for each `slice`).
+        group_indices
+            .iter()
+            .zip(input_group_values.iter())
+            .for_each(|(&group_index, values_opt)| {
+                if let Some(values) = values_opt {
+                    let values = values.as_primitive::<T>();
+                    self.group_values[group_index].extend(values.values().iter());
+                }
+            });
+
+        Ok(())
+    }
+
+    fn state(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>> {
+        // Emit values
+        let emit_group_values = emit_to.take_needed(&mut self.group_values);
+
+        // Build offsets
+        let mut offsets = Vec::with_capacity(self.group_values.len() + 1);
+        offsets.push(0);
+        let mut cur_len = 0;
+        for group_value in &emit_group_values {
+            cur_len += group_value.len() as i32;
```

Comment:
Do we need to check that this doesn't overflow an i32? I think it might be safer to do the following (or first check that the total number of items across all groups isn't more than 2B:

```suggestion
            cur_len = cur_len.checked_add(group_value.len()).unwrap();
```

Reply:
I test it in my local, found `check_add` is not obviously slower than `+`.
It is good to use `check_add + new_unchecked` here.

Reply:
I think of a cornercase, it may be possible `group_value.len() > i32::MAX`, and when cast it to i32, it will get `-x`.
Maybe we still need following check in `OffsetBuffer::new` for safety?
```
       assert!(
            buffer.windows(2).all(|w| w[0] <= w[1]),
            "offsets must be monotonically increasing"
        );
```

Reply:
> I think of a cornercase, it may be possible group_value.len() > i32::MAX, and when cast it to i32, it will get -x.
Maybe we still need following check in OffsetBuffer::new for safety?

Maybe we can leave it as is for now

Another thing to do is to check on `group_value.len()` conversion

like
```rust
let len: i32 = group_value.len().try_into().unwrap()
```

or something

Reply:
I left a todo about this, and just keep using `OffsetBuffer::new` currently.

Reply:
makes sense to me -- this PR is already better than what is on master so we can keep optimizing if needed afterwards